frontend/src/d3/d3.ts

```diff
@@ -0,0 +1,191 @@
+  forceY,
+  select,
+  Selection,
+  Simulation,
+  SimulationNodeDatum,
+} from 'd3';
+import { GraphEdge, GraphNode } from '../types/ontologyTypes';
+
+export const createForceSimulation = (
+  nodes: GraphNode[],
+  links: GraphEdge[],
+): Simulation<GraphNode, GraphEdge> =>
+  forceSimulation(nodes)
+    .force('charge', forceManyBody().strength(-999))
+    .force('center', forceCenter(500 / 2, 500 / 2))
+    .force('collide', forceCollide().radius(5)) // set radius dynamically
+    .force(
+      'x',
+      forceX((node) => node.x!),
+    )
+    .force(
+      'y',
+      forceY((node) => node.y!),
+    )
+    .force(
+      'link',
+      forceLink()
+        .id((node: SimulationNodeDatum) => (node as GraphNode).id)
+        .links(links)
+        .distance((link) =>
+          Math.min(links.filter((l) => l.source === link.target).length * 10 + 30, 400),
+        )
+        .strength(1),
+    )
+    .alpha(1);
+
+export const resetSimulation = (
+  simulation: Simulation<GraphNode, GraphEdge>,
+  nodes: GraphNode[],
+  links: GraphEdge[],
+) => {
+  simulation.nodes(nodes);
+  const linkForce = simulation.force<ForceLink<GraphNode, GraphEdge>>('link');
+  if (!linkForce) return;
+  linkForce.links(links);
+  simulation.force('collide');
+  simulation.alpha(1).restart();
+};
+
+export const drawLinks = (
+  svg: Selection<SVGSVGElement | null, unknown, null, undefined>,
+  links: GraphEdge[],
+  linkClassName: string,
+  fillColor: string,
+  strokeColor: string,
+  strokeWidth: number,
+) => {
+  svg
+    .selectAll(linkClassName)
+    .data(links)
+    .join('line')
+    .attr('class', linkClassName.substring(1)) // remove . before class name
+    .attr('fill', fillColor)
+    .attr('stroke', strokeColor)
+    .attr('stroke-width', strokeWidth);
+};
+
+export const updateLinkPositions = (
+  svg: Selection<SVGSVGElement | null, unknown, null, undefined>,
+  links: GraphEdge[],
+  linkClassName: string,
+) => {
+  svg
+    .selectAll(linkClassName)
+    .data(links)
+    .join('line')
+    .attr('x1', (link: any) => link.source.x)
+    .attr('y1', (link: any) => link.source.y)
+    .attr('x2', (link: any) => link.target.x)
+    .attr('y2', (link: any) => link.target.y);
+};
+
+export const drawNodes = (
+  svg: Selection<SVGSVGElement | null, unknown, null, undefined>,
+  nodes: GraphNode[],
+  nodeClassName: string,
+  radius: number,
+  fillColor: string,
+  highlightColor: string,
+  onClick: (node: GraphNode) => void,
+) => {
+  svg
+    .selectAll(nodeClassName)
+    .data(nodes)
+    .join((enter) => enter.append('circle').attr('cx', 100).attr('cy', 100))
+    .attr('class', nodeClassName.substring(1)) // remove . from class name
+    .attr('r', radius)
+    .attr('fill', fillColor)
+    // eslint-disable-next-line func-names
+    .on('mouseover', function () {
+      select(this)
+        .attr('fill', highlightColor)
+        .transition('500')
+        .attr('r', radius * 2);
+    })
+    // eslint-disable-next-line func-names
```

Comment:
comment needed?

Reply:
Yes as it allows ESlint to accept using `function() {}` instead of `() => {}` as it is required in this usecase